### PR TITLE
Protocol redefinition case and defmulti case

### PR DESCRIPTION
**Pull request by ericnormand**
Redefining a protocol loses all of the types that abide by it. If you redefine the protocol, which could be in file A, you need to rerun all `deftypes`, `defrecords`, `extend-types`, and `extend-protocols` that refer to it, which could be in any number of files B, C, ..., Z.

Re-evaluating a `defmulti` has no effect if the name is unchanged. If you redefine the dispatch function, it will not change the original multi, and will thus give you incorrect results compared to removing the `defmulti`'s Var and compiling it again.

### CODE REVIEW

File: src/clojure_repl_sufficient_p/defmulti.clj [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+(defmethod color-change :octopus
+  [octopus color]
+  (assoc octopus :color color))
+
+(def pre-iteration-result (color-change {:animal :octopus :color :white} :red))
+
+(defmulti color-change (fn [x & _] (:or (:animal x)
+                                        (:plant  x))))
+
```
Comment (aredington) [inline src/clojure_repl_sufficient_p/defmulti.clj:17]: For this to be a good case, we should have a `(defmethod color-change :hydrangea ...)` case at the minimum given the expectations of the test. 

As it turns out, I scrapped a multimethod namespace because it was surprising me with how well it behaved. Which is not to say we can't find edge cases, but we do need to be crisp about them actually being Clojure's problems. :)

**Review by aredington — CHANGES_REQUESTED** [review]
I think exploring more of the warts around defprotocol is a great idea. Let's attack the entire space:

- defrecord implicitly implementing a protocol, and the protocol is redefined
- deftype implicitly implementing a protocol, and the protocol is redefined

- defrecord implicitly implementing a protocol, and the record is redefined
- deftype implicitly implementing a protocol, and the type is redefined

- extend-protocol on extant record and protocol, redefine protocol
- extend-protocol on extant record and protocol, redefine record
- extend-protocol on extant type and protocol, redefine protocol
- extend-protocol on extant type and protocol, redefine type
- extend-type on same 4-branches cartesian product as prior

A lot of these things WORK. That's good, we want our programming language to work. Discussing all of this before merging isn't to dump a load of work in your lap, let's divide and conquer up this description of work. But let's do something more organized than `defprotocol.clj` `defprotocol2.clj` ... etc

File: src/clojure_repl_sufficient_p/defmulti.clj (hunk 1 above)
Comment (ericnormand) [reply src/clojure_repl_sufficient_p/defmulti.clj:17]: Agreed. It was an oversight.

**Comment by ericnormand** [comment]
Okay, I've covered 14 scenarios dealing with protocols.

**Review by aredington — APPROVED** [review]

**Comment by aredington** [comment]
I'm going to add some editorial notes and also some passing test cases. Particularly because changing the default of a defmulti DOES work correctly, so changing the dispatch fn and it not working highlights how this particular case really causes bad interface consistency. (That is, the interface Clojure offers its developer users.)
